### ISSUE TITLE
Allow for TLS configurations for Routes

### ISSUE THREAD

**Issue by fahlmant** [openshift/operator-custom-metrics#15]
Currently there is no way to configure a secure Route. WithRouteTLS() or some equivalent should be created. Allows for either Pass through or edge termination and the associated options.

**Comment by varshaprasad96** [comment]
@fahlmant : Can I work on this, if this is still undone.
If so, is this the approach which could be followed:
If the user provides the path for TLS Certificate and TLS key we could use the ListenAndServeTLS function (similar to what is provided by SDK), to initialise the connection?

**Comment by fahlmant** [comment]
Yes, basically support for Route TLS based on this section of the OpenShift documentation https://docs.openshift.com/container-platform/3.11/architecture/networking/routes.html#secured-routes

**Comment by clcollins** [comment]
I'm going to take a crack at this.

It's kind of ugly dropping the PEM key into the route object (edge termination), but that looks like it's dependent on an upstream OCP feature to migrate the key/cert to a secret, and it's not been addressed for a while: https://github.com/openshift/origin/issues/2162

@fahlmant Is there a desire to support end-to-end encryption and terminate TLS at the pod (passthrough encryption, ie, @varshaprasad96's ListenAndServeTLS comment)?  That would be a safer way to accomplish this, and would allow us to store the PEM files in secrets rather than plain text.

**Comment by clcollins** [comment]
Actually, this looks like there may have been an attempt already: https://github.com/openshift/operator-custom-metrics/pull/20

I'll review that and provide feedback.

Thanks @varshaprasad96!

**Comment by fahlmant** [comment]
@clcollins Passthrough would be ideal as you said because it's safer and ensures the PEMs are in Secrets. Edge termination is less attractive as an option in terms of security, but I threw it out there since it would simplify the code needed in the library a bit. We can explore that path if a use case of feature request comes up.
